src/web/app/pages-session/session-submission-page/session-submission-page.component.spec.ts

```diff
@@ -1,24 +1,565 @@
+import { AuthService } from '../../../services/auth.service';
+import { FeedbackQuestionsService } from '../../../services/feedback-questions.service';
+import { FeedbackResponseCommentService } from '../../../services/feedback-response-comment.service';
+import { FeedbackResponsesService } from '../../../services/feedback-responses.service';
+import { FeedbackSessionsService } from '../../../services/feedback-sessions.service';
+import { InstructorService } from '../../../services/instructor.service';
+import { NavigationService } from '../../../services/navigation.service';
+import { SimpleModalService } from '../../../services/simple-modal.service';
+import { StudentService } from '../../../services/student.service';
+import {
+  AuthInfo,
+  CommentVisibilityType,
+  FeedbackConstantSumQuestionDetails,
+  FeedbackConstantSumResponseDetails,
+  FeedbackContributionQuestionDetails,
+  FeedbackContributionResponseDetails,
+  FeedbackMcqQuestionDetails,
+  FeedbackMcqResponseDetails,
+  FeedbackMsqQuestionDetails,
+  FeedbackMsqResponseDetails,
+  FeedbackNumericalScaleQuestionDetails,
+  FeedbackNumericalScaleResponseDetails,
+  FeedbackParticipantType,
+  FeedbackQuestionRecipients,
+  FeedbackQuestions,
+  FeedbackQuestionType,
+  FeedbackRankOptionsQuestionDetails,
+  FeedbackRankOptionsResponseDetails,
+  FeedbackRankRecipientsQuestionDetails,
+  FeedbackResponse,
+  FeedbackResponseComment,
+  FeedbackResponses,
+  FeedbackRubricQuestionDetails,
+  FeedbackRubricResponseDetails,
+  FeedbackSession,
+  FeedbackSessionPublishStatus,
+  FeedbackSessionSubmissionStatus,
+  FeedbackTextQuestionDetails,
+  FeedbackTextResponseDetails,
+  FeedbackVisibilityType,
+  Instructor,
+  JoinState,
+  NumberOfEntitiesToGiveFeedbackToSetting,
+  RegkeyValidity,
+  ResponseVisibleSetting,
+  SessionVisibleSetting,
+  Student,
+} from '../../../types/api-output';
+import { Intent } from '../../../types/api-request';
 import { AjaxLoadingModule } from '../../components/ajax-loading/ajax-loading.module';
 import { LoadingRetryModule } from '../../components/loading-retry/loading-retry.module';
 import { LoadingSpinnerModule } from '../../components/loading-spinner/loading-spinner.module';
+import {
+  FeedbackResponseRecipientSubmissionFormModel,
+  QuestionSubmissionFormModel,
+} from '../../components/question-submission-form/question-submission-form-model';
 import {
   QuestionSubmissionFormModule,
 } from '../../components/question-submission-form/question-submission-form.module';
 import { TeammatesCommonModule } from '../../components/teammates-common/teammates-common.module';
+import { SavingCompleteModalComponent } from './saving-complete-modal/saving-complete-modal.component';
 import { SessionSubmissionPageComponent } from './session-submission-page.component';
+import Spy = jasmine.Spy;
 
 describe('SessionSubmissionPageComponent', () => {
+  const deepCopy: any = (obj: any) => JSON.parse(JSON.stringify(obj));
```

Comment:
Minor but you can use this instead:
```
  function deepCopy<T>(obj: T): T {
    return JSON.parse(JSON.stringify(obj));
  }
```
Read more about generics here: https://www.typescriptlang.org/docs/handbook/2/generics.html

> While using any is certainly generic in that it will cause the function to accept any and all types for the type of arg, we actually are losing the information about what that type was when the function returns. If we passed in a number, the only information we have is that any type could be returned.

Reply:
Been trying to get it to work with type variables, but lint doesn't allow for non-arrow functions, and I can't find a way to use them on arrow functions without errors. `const deepCopy = <T>(obj: T) => ...` needs a return type while trying things like `const deepCopy: T = <T>(obj: T) => ...` doesn't seem to work?

Reply:
See if this works https://stackoverflow.com/a/45576880

Reply:
None of those worked :sweat_smile: but I think this new one is ok